### PR TITLE
add variable sync pool

### PR DESCRIPTION
**Pull request by carolove**
give variable before give ctx

add variable test

fix small pool code

reset variable when put pool

small fix at variable pool

small fix about variable pool

### Issues associated with this PR

Your PR should present related issues you want to solve.

### Solutions
You should show your solutions about the issues in your PR, including the overall solutions, details and the changes. At this time, Chinese is allowed to describe these.

### UT result
Unit Test is needed if the code is changed, your unit test should cover boundary cases, corner cases, and some exceptional cases. And you need to show the UT result.

### Benchmark
If your code involves the processing of every request, you should give the Benchmark Result.

### Code Style
+ Make sure `Goimports` has run
+ Show `Golint` result

### CODE REVIEW

**Comment by taoyuanyuan** [comment]
add Performance banchmark

**Comment by taoyuanyuan** [comment]
#1682

File: pkg/variable/factory.go [hunk 1]
```diff
@@ -123,9 +128,35 @@ func RegisterPrefixVariable(prefix string, variable Variable) error {
...
+func (ctx variableBufferCtx) New() interface{} {
+	return &variableBuffer{
+		vars: make([]IndexedValue, len(indexedVariables)),
+	}
+}
+
+func (ctx variableBufferCtx) Reset(i interface{}) {
+	buf, _ := i.(*variableBuffer)
+	for idx := range buf.vars {
+		buf.vars[idx].data = ""
```
Comment (taoyuanyuan) [inline pkg/variable/factory.go:146]: buf.vars[idx] = IndexedValue{}

File: pkg/variable/factory.go [hunk 2]
```diff
@@ -23,10 +23,15 @@ import (
 	"strings"
 	"sync"
 
+	"mosn.io/mosn/pkg/buffer"
 	mosnctx "mosn.io/mosn/pkg/context"
 	"mosn.io/mosn/pkg/types"
 )
 
+func init() {
```
Comment (taoyuanyuan) [inline pkg/variable/factory.go:31]: move to a  new file named buffer.go

**Comment by nejisama** [comment]
ping @carolove

**Pull request closed without merging**